nntrainer/src/fc_layer.cpp

```diff
@@ -253,57 +164,28 @@ Tensor FullyConnectedLayer::backwarding(Tensor derivative, int iteration) {
   if (!last_layer) {
     djdb = derivative.multiply(hidden.apply(activation_prime));
   } else {
-    if (activation_type == ACT_SOFTMAX)
-      y = hidden.apply(softmax);
-    else
-      y = hidden.apply(activation);
-    float ll = opt.getLearningRate();
-    if (opt.getDecaySteps() != -1) {
-      ll = ll * pow(opt.getDecayRate(), (iteration / opt.getDecaySteps()));
-    }
-
     switch (cost) {
-    case COST_MSR: {
-      Tensor sub = y2.subtract(y);
-      Tensor l = (sub.multiply(sub)).sum_by_batch().multiply(0.5);
-
-      updateLoss(l);
-
+    case COST_MSR:
       if (activation_type == ACT_SOFTMAX) {
-        djdb = y.subtract(y2).multiply(y.apply(softmaxPrime));
+        y = hidden.apply(softmax);
+        djdb = derivative.multiply(hidden.apply(softmaxPrime));
       } else {
-        djdb = y.subtract(y2).multiply(hidden.apply(activation_prime));
-      }
-    } break;
-    case COST_ENTROPY: {
-      Tensor l;
-      if (activation_type == ACT_SIGMOID) {
-        djdb = y.subtract(y2).multiply(1.0 / y.getWidth());
-        l = (y2.multiply(y.apply(logFloat))
-               .add((y2.multiply(-1.0).add(1.0))
-                      .multiply((y.multiply(-1.0).add(1.0)).apply(logFloat))))
-              .multiply(-1.0 / (y2.getWidth()))
-              .sum_by_batch();
-      } else if (activation_type == ACT_SOFTMAX) {
-        djdb = y.subtract(y2).multiply(1.0 / y.getWidth());
-        l = (y2.multiply(y.apply(logFloat)))
-              .multiply(-1.0 / (y2.getWidth()))
-              .sum_by_batch();
-      } else {
-        ml_loge("Only support sigmoid & softmax for cross entropy loss");
-        exit(0);
+        djdb = derivative.multiply(hidden.apply(activation_prime));
       }
+      break;
+
+    case COST_ENTROPY:
+      djdb = derivative;
```

Comment:
Sorry for the late review. Could you explain why `djdb` is not updated when `COST_ENTROPY`?

Reply:
For `COST_ENTROPY`, the derivative with softmax has been optimized, where many terms cancel out and the final term which remains has been put into the `loss_layer` which is just diff of output and label.

Reply:
Right. 

I missed `if (!last_layer) { } else { }`. Thank you for the explanation :)